run/markdown-preview/editor/render.js

```diff
@@ -0,0 +1,63 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Import the Metadata API
+const gcpMetadata = require('gcp-metadata')
+const got = require('got');
+
+// renderRequest creates a new HTTP request with IAM ID Token credential.
+// This token is automatically handled by private Cloud Run (fully managed) and Cloud Functions.
+const renderRequest = async (service, markdown) => { 
+  // [START run_secure_request]
+  let token;
+
+  // Build the request to the Renderer receiving service.
+  const serviceRequestOptions = { 
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    body: JSON.stringify({markdown: markdown})
+  };
+
+  if (service.isAuthenticated) {
+    setTimeout( async () => {
+      try {
+      // Query the token with ?audience as the service URL.
+      const metadataServerTokenPath = `service-accounts/default/identity?audience=${service.url}`;
+      // Fetch the token and then add it to the request header.
+      token = await gcpMetadata.instance(metadataServerTokenPath);
+      serviceRequestOptions.headers['Authorization'] = 'bearer ' + token;
+      } catch(err) {
+        console.log('Metadata server could not respond to request ', err);
+        return err;
+      }
+    }, 2000);
+  };
+  // [END run_secure_request]
+
+  // [START run_secure_request_do]
+  try {
+    // serviceRequest converts the Markdown plaintext to HTML.
+    const serviceRequest = await got(service.url, serviceRequestOptions);
+    const serviceResponse = serviceRequest.body;
+    return serviceResponse;
+  } catch (err) { 
+    console.log('Renderer service could not respond to request ', err);
+    return err;
+  };
+  // [END run_secure_request_do]
```

Comment:
This region tag should probably go around where you use this function.

Reply:
I think the tag is in the correct place corresponding to its place in the Go code, but I could use clarification if that's not the case.

Reply:
run_secure_request wraps how to make the token request. run_secure_request_do is utilizing that function.

Reply:
Isn't the 'got' request to the Renderer service the thing that is utilizing that function? As opposed to the Renderer service itself, which isn't using the token but runs as a result of the token being used?

Reply:
Note that you don't need to implement the `_do` region tag unless you need to split the creation of the HTTP request from sending that request. In the case of Go this makes sense AND makes it unit testable. However, a different division may make more sense:

For example, just use `run_secure_request` and in that have one function to mint the token and another to use it and send an HTTP request.

Reply:
@kelsk region tags need to be updated

Reply:
Updated - the token minting and HTTP request are in separate blocks, but within the single region tag `run_secure_request`.